Amplify/Categories/DataStore/Model/Temporal/Temporal+Cache.swift

```diff
@@ -0,0 +1,172 @@
+    private static var formatterCache: [String: DateFormatter] = [:]
+
+    @usableFromInline
+    /// The `Calendar` used for date operations.
+    ///
+    /// `identifier` is `.iso8601`
+    /// `timeZome` is `.utc` a.k.a. `TimeZone(abbreviation: "UTC")`
+    internal static let iso8601Calendar: Calendar = {
+        var calendar = Calendar(identifier: .iso8601)
+        calendar.timeZone = .utc
+        return calendar
+    }()
+
+    /// Lock to ensure exclusive access
+    private static var lock = os_unfair_lock_s()
+
+    /// Internal helper function to retrieve and/or create `DateFormatter`s
+    /// - Parameters:
+    ///   - format: The `DateFormatter().dateFormat`
+    ///   - timeZone: The `DateFormatter().timeZone`
+    /// - Returns: A `DateFormatter`
+    internal static func formatter(
+        for format: String,
+        in timeZone: TimeZone
+    ) -> DateFormatter {
+        defer { os_unfair_lock_unlock(&lock) }
+
+        // lock before read from cache
+        os_unfair_lock_lock(&lock)
+
+        // If the formatter is already in the cache and
+        // the time zones match, we return it rather than
+        // creating a new one.
+        if let formatter = formatterCache[format],
+           formatter.timeZone == timeZone {
+                return formatter
+            // defer takes care of unlock
+        }
+
+        // If:
+        // - There's not another reference to this formatter
+        // - It's already in the cache
+        // - The formatter's time zone doesn't match the requested time zone
+        // Then:
+        // We can safely change the formatter's time zone and return it
+        if isKnownUniquelyReferenced(&formatterCache[format]),
+           let formatter = formatterCache[format],
+           formatter.timeZone != timeZone {
+            formatter.timeZone = timeZone
+            return formatter
+            // defer takes care of unlock
+        }
+
+        // We're about to create a new formatter,
+        // so let's make sure we're being responsible
+        // about the amount of formatters we're caching.
+        // If we already have at least 15 formatters in the cache,
+        // we're going to evict one using a
+        // random replacement (RR) eviction policy
+        if formatterCache.count >= 15 {
+            // From the `randomElement()` documentation.
+            // "A random element from the collection. If the collection is empty, the method returns nil."
+            // Since we just confirmed the cache isn't empty, and we're guaranteeing exclusive access
+            // through the lock, we can safely force unwrap here.
+            let keyToEvict = formatterCache.randomElement()!.key
+            formatterCache[keyToEvict] = nil
+            // This can likely be improved at a later time by
+            // using a LFU, or potentially LRU, eviction policy.
+        }
+
+        // unlock if no early return
+        os_unfair_lock_unlock(&lock)
```

Comment:
This is to let other waiting calls to acquire the lock and move forward? And below code is expensive.

Reply:
That's correct. It would also be fine to keep this locked while instantiating the new `DateFormatter`. Happy to make that change if there's a good reason to.

Reply:
Great catch and callout, thanks! That's almost certainly what's causing the intermittent DataStore test failures. Just pushed [b2062c2](https://github.com/aws-amplify/amplify-ios/pull/1760/commits/b2062c2cc8709fe839982b5209df9d9a2d5bea78) after tests passed locally with the change.